doc: Add doc for Bluetooth mesh SAR Config models (#61886)
zephyrproject-rtos/zephyr: doc/connectivity/bluetooth/api/mesh/sar_cfg.rst

.. _bluetooth_mesh_sar_cfg:

Segmentation and reassembly (SAR)
#################################

Segmentation and reassembly provides a way of handling larger upper transport layer messages in a
mesh network, with a purpose of enhancing the Bluetooth mesh throughput. The segmentation and
reassembly mechanism is used by the lower transport layer.

The lower transport layer takes Upper Transport PDUs from the upper transport layer and sends them to the lower transport layer on a peer device.
If the Upper Transport PDU fits, it is sent in a single Lower Transport PDU.

> For longer packets, which do not fit into a single Lower Transport PDU, the lower transport layer performs segmentation, splitting the Upper Trasport

Reviewer: ```suggestion
For longer packets, which do not fit into a single Lower Transport PDU, the lower transport layer performs segmentation, splitting the Upper Transport
```

Delivery of an unsegmented message is not acknowledged.
But, an Upper Transport PDU that fits into one Lower Transport PDU can also be sent as a single-segment segmented message when acknowledgment
by the lower transport layer is required.
Set the ``send rel`` flag (see :c:struct:`bt_mesh_msg_ctx`) to use the reliable message transmission and acknowledge single-segment segmented messages.

The transport layer is able to transport up to 32 segments with its SAR mechanism, with a maximum message (PDU) size of 384 octets.
To configure number of segments and the message size for the Bluetooth mesh stack, use the following Kconfig options:

* :kconfig:option:`CONFIG_BT_MESH_SEG_BUFS` to set the number of segment buffers.
* :kconfig:option:`CONFIG_BT_MESH_RX_SEG_MAX` to set the maximum number of segments in an incoming message.
* :kconfig:option:`CONFIG_BT_MESH_TX_SEG_MAX` to set the maximum number of segments in an outcoming message.

Never more than one segmented message is transmitted at the time, but the Bluetooth mesh stack allows you to define buffers that will queue a defined

> number of ingoing or outgoing segmented messages.

Author: Right...
It's been a few ingoing and outcoming messages LOL Unbelievable!

Use Kconfig options :kconfig:option:`CONFIG_BT_MESH_TX_SEG_MSG_COUNT` and :kconfig:option:`CONFIG_BT_MESH_RX_SEG_MSG_COUNT` to set the maximum number

> of queued outgoing and incoming segmented messages respectively.

Author: :kconfig:option:`CONFIG_BT_MESH_TX_SEG_MSG_COUNT` defines outgoing messages, right? And that one is mentioned first in the sentence. That's why outgoing first, and then incoming. Would you like me to swap the Kconfig options to make your suggestion right?

When the last segment marked as unacknowledged is transmitted, the lower transport layer starts a retransmissions timer.
The initial value of the SAR Unicast Retransmissions timer depends on the value of the TTL field of the message.
If the TTL field value is greater than ``0``, the initial value for the timer is set according to the following formula::

[unicast retransmissions interval step + unicast retransmissions interval increment * (TTL - 1)]

If the TTL field value is ``0``, the initial value of the timer is set to the unicast retransmissions interval step.

The initial value of the SAR Multicast Retransmissions timer is set to the multicast retransmissions interval.

When the lower transport layer receives a segment of a segmented message, it starts an SAR Discard timer.
The discard timer tells how long the lower transport layer waits before discarding the segmented message the segment belongs to.
The initial value of the SAR Discard timer is the discard timeout value indicated by the `SAR Discard Timeout`_ state.

SAR Acknowledgment timer holds the time before a Segment Acknowledgment message is sent for a received segment.

> The initial value of the SAR Acknowledgment timer is calculated using the following formula::
> 
> [min(SegN + 0.5 , acknowledgment delay increment) * segment reception interval]

Reviewer: You should consider using actual math formulas as they render much more nicely.

```suggestion
The initial value of the SAR Acknowledgment timer is calculated using the following formula:

.. math::

   \min(\text{SegN} + 0.5, \text{acknowledgment delay increment}) \times \text{segment reception interval}
```

Alternatively, this simplified form works too

```suggestion
The initial value of the SAR Acknowledgment timer is calculated using the following formula:

.. math::

   min(SegN + 0.5 , acknowledgment~delay~increment) * segment~reception~interval
```

Finally, it's also possible to use the `:math:` role if you want the formula in-line with the text (``blah blah :math:`my formula` blah blah``).

Reviewer: FWIW I would recommend you use the math directive everywhere, not just in this particular instance.
Rendered doc available at https://builds.zephyrproject.io/zephyr/pr/61886/docs/connectivity/bluetooth/api/mesh/sar_cfg.html , in case you are not aware and/or are not testing locally.

Author: Thanks Benjamin :)
Yeah, I know how they render. Personally I prefer code block rendition over the one using math, it's just something stylistically off with that one for me, for some reason I cannot quite explain. That's why I've been holding back.
But, I'll give it a go and see if it starts growing on me LOL

Four counters are related to SAR behavior:

> * Unicast retransmissions count (see `SAR Unicast Retransmissions Count`_ state and `SAR Unicast Retransmissions Without Progress Count`_ state)

Reviewer 2: ```suggestion
* Two Unicast retransmissions counts (see `SAR Unicast Retransmissions Count`_ state and `SAR Unicast Retransmissions Without Progress Count`_ state)
```

The SAR Transmitter state is a composite state that controls the number and timing of transmissions of segmented messages.
It includes the following states:

* SAR Segment Interval Step
* SAR Unicast Retransmissions Count
* SAR Unicast Retransmissions Without Progress Count
* SAR Unicast Retransmissions Interval Step
* SAR Unicast Retransmissions Interval Increment
* SAR Multicast Retransmissions Count
* SAR Multicast Retransmissions Interval Step

The SAR Receiver state is a composite state that controls the number and timing of Segment Acknowledgment transmissions and the discarding of reassembly of a segmented message.
It includes the following states:

* SAR Segments Threshold
* SAR Discard Timeout
* SAR Acknowledgment Delay Increment
* SAR Acknowledgment Retransmissions Count
* SAR Receiver Segment Interval Step

SAR Segment Interval Step
=========================

SAR Segment Interval Step state holds a value that controls the interval between transmissions of segments of a segmented message.
Use the :kconfig:option:`CONFIG_BT_MESH_SAR_TX_SEG_INT_STEP` Kconfig option to set a value.
Segment transmission interval is then calculated using the following formula::

> (:kconfig:option:`CONFIG_BT_MESH_SAR_TX_SEG_INT_STEP` + 1) * 10 ms

Reviewer: you can't use the `:kconfig:option:` role within a code block as it doesn't render

<img width="640" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/128251/11627bd7-f151-4506-aa51-386afbe8748a">

Author: Of course, I forgot to think of that. Removed the :kconfig:option:``

The receiving lower transport layer on the other device reassembles the segments into a single Upper Transport PDU before passing it
up the stack. Delivery of a segmented message is acknowledged by the lower transport layer of the receiving node.

Delivery of an unsegmented message is not acknowledged.
But, an Upper Transport PDU that fits into one Lower Transport PDU can also be sent as a single-segment segmented message when acknowledgment
by the lower transport layer is required.
Set the ``send rel`` flag (see :c:struct:`bt_mesh_msg_ctx`) to use the reliable message transmission and acknowledge single-segment segmented messages.

The transport layer is able to transport up to 32 segments with its SAR mechanism, with a maximum message (PDU) size of 384 octets.
To configure number of segments and the message size for the Bluetooth mesh stack, use the following Kconfig options:

* :kconfig:option:`CONFIG_BT_MESH_SEG_BUFS` to set the number of segment buffers.
* :kconfig:option:`CONFIG_BT_MESH_RX_SEG_MAX` to set the maximum number of segments in an incoming message.
* :kconfig:option:`CONFIG_BT_MESH_TX_SEG_MAX` to set the maximum number of segments in an outcoming message.

> Never more than one segmented message is transmitted at the time, but the Bluetooth mesh stack allows you to define buffers that will queue a defined

Reviewer 3: Maybe something like this instead: "Only one segmented message is transmitted at a time, but the Bluetooth mesh stack allows you to **configure** buffers that can queue a specified number of incoming or outgoing segmented messages."

number of ingoing or outgoing segmented messages.
Use Kconfig options :kconfig:option:`CONFIG_BT_MESH_TX_SEG_MSG_COUNT` and :kconfig:option:`CONFIG_BT_MESH_RX_SEG_MSG_COUNT` to set the maximum number
of queued outgoing and incoming segmented messages respectively.

Incoming and outgoing segmented messages share the same segment buffer, which means that they allocate their segments from the same pool.

> Both incoming and outgoing messages allocate segments at the start of the transaction, but release them at the different time.

Reviewer 3: I am a bit confused by this sentence. "but release them at the different time", what is meant by this? Should it be "at different times"?

Author: at different times, for sure :)

While the outgoing segmented message releases its segments one by one as soon as they are acknowledged by the receiver, the incoming message will not
release the segments until the message is fully received.
Keep this in mind when defining the size of the buffers.

SAR does not impose extra overhead on the access layer payload per segment.

Segmentation and reassembly (SAR) Configuration models are foundation models defined by the

> `Bluetooth Mesh Protocol Specification version 1.1 <https://www.bluetooth.com/specifications/mesh-specifications/>`_.

Author: This is the first and only place the spec is referenced on this page.
I am not sure what page you mean, maybe mesh.rst?
I also see we are only mentioning v1.0.1 there, no 1.1.

> `Bluetooth Mesh Protocol Specification version 1.1 <https://www.bluetooth.com/specifications/mesh-specifications/>`_.

Reviewer 4: > This is the first and only place the spec is referenced on this page.
I am not sure what page you mean, maybe mesh.rst?

Yes, I meant mesh.rst. Otherwise, we need to add the reference to every model's doc page which is too much I think. 

> I also see we are only mentioning v1.0.1 there, no 1.1.

This should be fixed (not necessarily in your PR, but up to you).

These models support the configuration of SAR behavior for nodes in a mesh network.

The SAR Configuration models are:

.. toctree::
   :maxdepth: 1

   sar_cfg_cli
   sar_cfg_srv

Intervals, timers and counters
******************************

Transmission of segments is separated by a segment transmission interval (see the `SAR Segment Interval Step`_ state).

> Other time intervals and delays available for the segmentation and reassembly are:

Reviewer 3: ```suggestion
Other configurable time intervals and delays available for the segmentation and reassembly are:
```

* Interval between unicast retransmissions (see the states `SAR Unicast Retransmissions Interval Step`_ and `SAR Unicast Retransmissions Interval Increment`_).
* Interval between multicast retransmissions (see the `SAR Multicast Retransmissions Interval Step`_ state).
* Segment reception interval (see the `SAR Receiver Segment Interval Step`_ state).
* Acknowledgment delay increment (see the `SAR Acknowledgment Delay Increment`_ state)

When the last segment marked as unacknowledged is transmitted, the lower transport layer starts a retransmissions timer.
The initial value of the SAR Unicast Retransmissions timer depends on the value of the TTL field of the message.
If the TTL field value is greater than ``0``, the initial value for the timer is set according to the following formula::

[unicast retransmissions interval step + unicast retransmissions interval increment * (TTL - 1)]

If the TTL field value is ``0``, the initial value of the timer is set to the unicast retransmissions interval step.

The initial value of the SAR Multicast Retransmissions timer is set to the multicast retransmissions interval.

> When the lower transport layer receives a segment of a segmented message, it starts an SAR Discard timer.

Reviewer 3: ```suggestion
When the lower transport layer receives a message segment, it starts a SAR Discard timer.
```

Author: It depends on whether you pronounce SAR as [sɑr] or [es/eɪ/ar/]. Not sure which one should be used.

Reviewer 3: I believe [sar] is the most common but this is just nitpicking :)

The Kconfig option value is used to calculate the increment using the following formula::

(CONFIG_BT_MESH_SAR_TX_UNICAST_RETRANS_INT_INC + 1) * 25 ms

SAR Multicast Retransmissions Count
===================================

The state holds a value that controls the total number of retransmissions of a segmented message to a multicast address.
Use the Kconfig option :kconfig:option:`CONFIG_BT_MESH_SAR_TX_MULTICAST_RETRANS_COUNT` to set the value for this state.

SAR Multicast Retransmissions Interval Step
===========================================

This state holds a value that controls the interval between retransmissions of all segments in a segmented message to a multicast address.
The interval is measured in milliseconds.

Use the Kconfig option :kconfig:option:`CONFIG_BT_MESH_SAR_TX_MULTICAST_RETRANS_INT` to set a value that is used to calculate the interval
using the following formula::

(CONFIG_BT_MESH_SAR_TX_MULTICAST_RETRANS_INT + 1) * 25 ms

SAR Discard Timeout
===================

The value of this state defines the time since the last successfully received segment before giving up the reception of a segmented message.

> Use the Kconfig option :kconfig:option:`CONFIG_BT_MESH_SAR_RX_DISCARD_TIMEOUT` to set the value for this state.

Reviewer 2: What unit is this given in? (seconds/milliseconds)

Author: seconds

========================================

The SAR Acknowledgment Retransmissions Count state controls the number of retransmissions of Segment Acknowledgment messages sent by the lower transport layer.
It gives the total number of retranmissions of an acknowledgment message that the stack will additionally send when the size of segments in a
segmented message is above the :kconfig:option:`CONFIG_BT_MESH_SAR_RX_SEG_THRESHOLD` value.

Use the Kconfig option :kconfig:option:`CONFIG_BT_MESH_SAR_RX_ACK_RETRANS_COUNT` to set the value for this state.

The maximum number of transmissions of a Segment Acknowledgment message is :kconfig:option:`CONFIG_BT_MESH_SAR_RX_ACK_RETRANS_COUNT` + 1.

SAR Receiver Segment Interval Step
==================================

The SAR Receiver Segment Interval Step defines the segments reception interval step used for delaying the transmission of an acknowledgment message after receiving a new segment.
The interval is measured in milliseconds.

Use the Kconfig option :kconfig:option:`CONFIG_BT_MESH_SAR_RX_SEG_INT_STEP` to set a value that will be used to calculate the interval using the following formula::

(CONFIG_BT_MESH_SAR_RX_SEG_INT_STEP + 1) * 10 ms

> API reference

Reviewer 4: This can be removed. This API is not public.

The outgoing segmented message releases its segments one by one as soon as they are acknowledged by the receiver, while the incoming message releases
the segments first after the message is fully received.
Keep this in mind when defining the size of the buffers.

SAR does not impose extra overhead on the access layer payload per segment.

Intervals, timers and retransmission counters
*********************************************

The current stable stack implementation allows you to configure the following SAR behavior.

When sending a segmented message to a unicast address, the unacknowledged segments are repeated
the :kconfig:option:`CONFIG_BT_MESH_TX_SEG_RETRANS_COUNT` number of times before the transmission
is considered as failed. The same option configures a number of retransmissions to a group or virtual
address, but the transmission always succeedes after retransmitting all segments the configured
number of times.

The timeout between each retransmission to a unicast address and a group or virtual address is
configured by Kconfig options :kconfig:option:`CONFIG_BT_MESH_TX_SEG_RETRANS_TIMEOUT_UNICAST` and

> :kconfig:option:`CONFIG_BT_MESH_TX_SEG_RETRANS_TIMEOUT_GROUP` respectively.

Reviewer 2: This was a little bit hard to follow, "unicast address and a group or virtual address". First there is "and" then "or"? Which kconfig option is used for virtual address?

* :ref:`bluetooth_mesh_sar_cfg_srv`

The following SAR behavior applies regardless of the presence of a SAR Configuration Server on a node.

Transmission of segments is separated by a segment transmission interval (see the `SAR Segment Interval Step`_ state).
Other configurable time intervals and delays available for the segmentation and reassembly are:

* Interval between unicast retransmissions (see the states `SAR Unicast Retransmissions Interval Step`_ and `SAR Unicast Retransmissions Interval Increment`_).
* Interval between multicast retransmissions (see the `SAR Multicast Retransmissions Interval Step`_ state).
* Segment reception interval (see the `SAR Receiver Segment Interval Step`_ state).
* Acknowledgment delay increment (see the `SAR Acknowledgment Delay Increment`_ state).

When the last segment marked as unacknowledged is transmitted, the lower transport layer starts a retransmissions timer.
The initial value of the SAR Unicast Retransmissions timer depends on the value of the TTL field of the message.
If the TTL field value is greater than ``0``, the initial value for the timer is set according to the following formula:

.. math::

>    [unicast~retransmissions~interval~step + unicast~retransmissions~interval~increment \times (TTL - 1)]

Reviewer 4: Minor: do we need square brackets here?

Reviewer 4: Yes

.. _bluetooth_mesh_sar_cfg:

Segmentation and reassembly (SAR)
#################################

Segmentation and reassembly is a mechanism used by the lower transport layer.
It provides a way of handling larger upper transport layer messages in a mesh network, with a purpose of enhancing the Bluetooth mesh throughput.

> The lower transport layer takes PDUs from the upper transport layer and sends them to the lower transport layer on a peer device.

Reviewer 2: ```suggestion
The lower transport layer takes Upper Transport PDUs from the upper transport layer and sends them to the lower transport layer on a peer device.
```

> For longer packets, which do not fit into a single Lower Transport PDU, the lower transport layer performs segmentation, splitting the PDU
> into multiple Lower Transport PDUs.

Reviewer 2: ```suggestion
For longer packets, which do not fit into a single Lower Transport PDU, the lower transport layer performs segmentation, splitting the Upper Transport PDU
into multiple segments.
```

> The receiving lower transport layer on the other device reassembles the segments into a single upper transport layer PDU before passing it

Reviewer 2: ```suggestion
The receiving lower transport layer on the other device reassembles the segments into a single Upper Transport PDU before passing it
```

SAR does not impose extra overhead on the access layer payload per segment.

Segmentation and reassembly (SAR) Configuration models are foundation models defined by the Bluetooth Mesh Protocol Specification version 1.1.
These models support the configuration of SAR behavior for nodes in a mesh network.

The SAR Configuration models are:

.. toctree::
   :maxdepth: 1

   sar_cfg_srv
   sar_cfg_cli

Segmentation
************

Segmentation is performed by the lower transport layer of the transmitting node.
The lower transport layer checks if an Upper Transport PDU fits into a single Lower Transport PDU.

If the Upper Transport PDU fits, it is sent in a single Lower Transport PDU.
An Upper Transport PDU that fits into one Lower Transport PDU can also be sent as a single-segment segmented message when acknowledgment
by the lower transport layer is required.

If the Upper Transport PDU doesn’t fit, the lower transport layer divides the Upper Transport PDU into two or more Lower Transport PDUs.
If a segmented Upper Transport PDU is tagged with additional metadata, each Lower Transport PDU of the segmented Upper Transport PDU is
tagged with the same metadata.

> Delivery of a segmented message is acknowledged by the lower transport layer of the receiving node.

Reviewer 3: ```suggestion
Delivery of a segmented message is acknowledged by the lower transport layer of the receiving node,  while an unsegmented message delivery isn't acknowledged.
```

> Delivery of an unsegmented message is not acknowledged.

Reviewer 3: ```suggestion
```

   All Lower Transport PDUs from the same Upper Transport PDU are sent using the same IV Index from the common ``SeqAuth``.

SeqZero
   The ``SeqZero`` field is included in the segmented message and Segment Acknowledgment message to identify the Upper Transport PDU.
   The least significant 13 bits of the ``SeqAuth`` value constitute the ``SeqZero`` field value.

Transmission
============

Never more than one segmented message is transmitted to the same destination.
The segments are sent in order of ``SegO`` field value, starting with the lowest value.
Transmission of segments is separated by a segment transmission interval (value indicated by the SAR Segment Interval Step state).

It is not possible to send a segmented message when the ``SEQ`` field value is 8192 greater than the ``SeqAuth`` value.
If a segmented message has not been acknowledged by the time that the ``SEQ`` field gets this value, the transmission of the Upper
Transport PDU is canceled and the Lower Transport PDUs are delivered to the network layer.

Segments can be sent to a unicast address or to a group or a virtual address.

> 

Reviewer 2: Is it possible to merge the "Transmitting to a ...." into one and just explain what it unique for unicast address, as there is a lot of repeating text. It makes it less to read and maintain. Same for retransmition.
